engines/wage/combat.cpp

```diff
@@ -0,0 +1,914 @@
+
+	appendText(buf);
+
+	_world->move(obj, npc);
+}
+
+int WageEngine::getValidMoveDirections(Chr *npc) {
+	int directions = 0;
+	Scene *currentScene = npc->_currentScene;
+	for (int dir = 0; dir < 4; dir++) {
+		if (!currentScene->_blocked[dir]) {
+			int destX = currentScene->_worldX + directionsX[dir];
+			int destY = currentScene->_worldY + directionsY[dir];
+
+			Scene *scene = _world->getSceneAt(destX, destY);
+
+			if (scene != NULL && scene->_chrs.empty()) {
+				directions |= (1 << dir);
+			}
+		}
+	}
+
+	return directions;
+}
+
+void WageEngine::regen() {
+	Chr *player = _world->_player;
+	int curHp = player->_context._statVariables[PHYS_HIT_CUR];
+	int maxHp = player->_context._statVariables[PHYS_HIT_BAS];
+	int delta = maxHp - curHp;
+
+	if (delta > 0) {
+		int bonus = (int)(delta / (8 + _rnd->getRandomNumber(2)));
+		player->_context._statVariables[PHYS_HIT_CUR] += bonus;
+	}
+}
+
+void WageEngine::takeObj(Obj *obj) {
+	if ((int)_world->_player->_inventory.size() >= _world->_player->_maximumCarriedObjects) {
+		appendText("Your pack is full, you must drop something.");
+	} else {
+		char buf[256];
+
+		_world->move(obj, _world->_player);
+		int type = _world->_player->wearObjIfPossible(obj);
+		if (type == Chr::HEAD_ARMOR) {
+			snprintf(buf, 256, "You are now wearing the %s.", obj->_name.c_str());
+			appendText(buf);
+		} else if (type == Chr::BODY_ARMOR) {
+			snprintf(buf, 256, "You are now wearing the %s.", obj->_name.c_str());
+			appendText(buf);
+		} else if (type == Chr::SHIELD_ARMOR) {
+			snprintf(buf, 256, "You are now wearing the %s.", obj->_name.c_str());
+			appendText(buf);
+		} else if (type == Chr::MAGIC_ARMOR) {
+			snprintf(buf, 256, "You are now wearing the %s.", obj->_name.c_str());
+			appendText(buf);
+		} else {
+			snprintf(buf, 256, "You now have the %s.", obj->_name.c_str());
+			appendText(buf);
+		}
+		appendText(obj->_clickMessage.c_str());
+	}
+}
+
+bool WageEngine::handleMoveCommand(Directions dir, const char *dirName) {
+	Scene *playerScene = _world->_player->_currentScene;
+	const char *msg = playerScene->_messages[dir].c_str();
+
+	if (!playerScene->_blocked[dir]) {
+		int destX = playerScene->_worldX + directionsX[dir];
+		int destY = playerScene->_worldY + directionsY[dir];
+
+		Scene *scene = _world->getSceneAt(destX, destY);
+
+		if (scene != NULL) {
+			if (strlen(msg) > 0) {
+				appendText(msg);
+			}
+			_world->move(_world->_player, scene);
+			return true;
+		}
+	}
+	if (strlen(msg) > 0) {
+		appendText(msg);
+	} else {
+		Common::String txt("You can't go ");
+		txt += dirName;
+		txt += ".";
+		appendText(txt.c_str());
+	}
+
+	return true;
+}
+
+bool WageEngine::handleLookCommand() {
+	appendText(_world->_player->_currentScene->_text.c_str());
+
+	Common::String *items = getGroundItemsList(_world->_player->_currentScene);
+	if (items != NULL) {
+		appendText(items->c_str());
+
+		delete items;
+	}
+
+	return true;
+}
+
+Common::String *WageEngine::getGroundItemsList(Scene *scene) {
```

Comment:
It looks like this is used in two places. One right above, which checks for `NULL` as possible return value, and one in `WageEngine::attackHit` which does not care about it possibly returning `NULL`. Since it looks like when this returns non-`NULL` is always empty it might be more sensible to return a `Common::String` and perform `empty` checks. That also avoids manual memory management.